src/lib/provable-types/merkle-list.ts

```diff
@@ -0,0 +1,495 @@
+import { Bool, Field } from '../core.js';
+import { Provable } from '../provable.js';
+import { Struct, Unconstrained } from '../circuit_value.js';
+import { assert } from '../gadgets/common.js';
+import { provableFromClass } from '../../bindings/lib/provable-snarky.js';
+import { Poseidon, packToFields, ProvableHashable } from '../hash.js';
+
+export {
+  MerkleListBase,
+  MerkleList,
+  MerkleListIteratorBase,
+  MerkleListIterator,
+  WithHash,
+  emptyHash,
+  genericHash,
+  merkleListHash,
+  withHashes,
+};
+
+// common base types for both MerkleList and MerkleListIterator
+
+const emptyHash = Field(0);
+
+type WithHash<T> = { previousHash: Field; element: T };
+
+function WithHash<T>(type: ProvableHashable<T>): ProvableHashable<WithHash<T>> {
+  return Struct({ previousHash: Field, element: type });
+}
+
+/**
+ * Common base type for {@link MerkleList} and {@link MerkleListIterator}
+ */
+type MerkleListBase<T> = {
+  hash: Field;
+  data: Unconstrained<WithHash<T>[]>;
+};
+
+function MerkleListBase<T>(): ProvableHashable<MerkleListBase<T>> {
+  return class extends Struct({ hash: Field, data: Unconstrained.provable }) {
+    static empty(): MerkleListBase<T> {
+      return { hash: emptyHash, data: Unconstrained.from([]) };
+    }
+  };
+}
+
+// merkle list
+
+/**
+ * Dynamic-length list which is represented as a single hash
+ *
+ * Supported operations are {@link push()} and {@link pop()} and some variants thereof.
+ *
+ *
+ * A Merkle list is generic over its element types, so before using it you must create a subclass for your element type:
+ *
+ * ```ts
+ * class MyList extends MerkleList.create(MyType) {}
+ *
+ * // now use it
+ * let list = MyList.empty();
+ *
+ * list.push(new MyType(...));
+ *
+ * let element = list.pop();
+ * ```
+ *
+ * Internal detail: `push()` adds elements to the _start_ of the internal array and `pop()` removes them from the start.
```

Comment:
what about naming the methods  `unshift` and `shift` for `push` and `pop` respectively so that the method the overall `MerkleList` structure follows the native JavaScript array more closely?

Reply:
Hm.. you have a point but what gives me pause is that the names `shift()` / `unshift()` are probably less well-known and recognizable for developers not intimately familiar with the JS stdlib